GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/MultithreadedReadWriteTests.cs

```diff
@@ -133,32 +164,69 @@ public void CanReadWriteAFileInParallel(FileSystemRunner fileSystem)
                 while ((DateTime.Now - start).TotalSeconds < 2.5)
                 {
                     string newChar = r.Next(10).ToString();
-                    fileSystem.AppendAllText(virtualPath, newChar);
-                    fileContents[0].Append(newChar);
+                    try
+                    {
+                        fileSystem.AppendAllText(virtualPath, newChar);
+                        fileContents[0].Append(newChar);
+                    }
+                    catch (IOException) when (Marshal.GetLastWin32Error() == this.nativeEWouldBlock && this.nativeEWouldBlock > 0)
+                    {
+                        // ignore EAGAIN
+                    }
+                    catch (Exception e)
+                    {
+                        writeException = e;
+                    }
                     Thread.Yield();
                 }
 
                 keepRunning = false;
             });
 
+            Exception readException = null;
+
             // Readers
             for (int i = 1; i < threads.Length; ++i)
             {
                 int myIndex = i;
                 fileContents[i] = new StringBuilder();
                 threads[i] = new Thread(() =>
                 {
-                    using (Stream readStream = File.Open(virtualPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
-                    using (StreamReader reader = new StreamReader(readStream, true))
+                    bool retry = true;
+                    while (retry)
                     {
-                        while (keepRunning)
+                        Stream readStream = null;
+                        try
                         {
-                            Thread.Yield();
-                            fileContents[myIndex].Append(reader.ReadToEnd());
+                            readStream = File.Open(virtualPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
+                            retry = false;

```

Comment:
Am I reading this right that this loop will repeat infinitely if `File.Open()` throws? What's the benefit there?

Reply:
(and that's the only way it will loop at all)

Reply:
Good catch and question -- the workaround here is to deal with `EAGAIN` from the `File.Open()`.  I think I'm missing some additional handling around the non-`EAGAIN` case, where we want to exit the loop.  Stay tuned and thanks!